### PR TITLE
Refactor BigQuery query API

### PR DESCRIPTION
**Pull request by jskeet**

### CODE REVIEW

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryClientImpl.Queries.cs [hunk 1]
```diff
@@ -167,33 +125,41 @@ public override BigQueryResults GetQueryResults(JobReference jobReference, GetQu
...
-        public override async Task<BigQueryJob> CreateQueryJobAsync(BigQueryCommand command, CreateQueryJobOptions options = null, CancellationToken cancellationToken = default(CancellationToken))
+        public override async Task<BigQueryJob> CreateQueryJobAsync(BigQueryCommand command, QueryOptions options = null, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(command, nameof(command));
             var query = new JobConfigurationQuery { UseLegacySql = false };
             command.PopulateJobConfigurationQuery(query);
             var job = await CreateInsertQueryJobRequest(query, options).ExecuteAsync(cancellationToken).ConfigureAwait(false);
             return new BigQueryJob(this, job);
-        }
+        }
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryClientImpl.Queries.cs]: Nit: unnecessary whitespace added.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs [hunk 2]
```diff
@@ -201,6 +189,22 @@ public BigQueryJob ThrowOnAnyError()
...
             _client.CancelJobAsync(Reference, options, cancellationToken);
 
+        private TableReference GetQueryTableReference()
+        {
+            if (Resource?.Configuration?.Query == null)
+            {
+                throw new InvalidOperationException("Job doesn't represent a query");
+            }
+            ThrowOnAnyError();
+            var reference = Resource?.Configuration?.Query?.DestinationTable;
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs]: Nit: since we already know that `Resource?.Configuration?.Query` is non-null, this can be `Resource.Configuration.Query.DestinationTable`. Although I think it would be better to cache `Resource?.Configuration?.Query` is a local above, check for null, and then use it here.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/ListRowsOptionsTest.cs [hunk 3]
```diff
@@ -21,15 +22,44 @@ namespace Google.Cloud.BigQuery.V2.Tests
...
+
+        [Fact]
+        public void ModifyRequest_StartIndex()
+        {
+            var options = new ListRowsOptions
+            {
+                PageSize = 25,
+                StartIndex = 10
+            };
+            ListRequest request = new ListRequest(new BigqueryService(), "project", "dataset", "table");
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/ListRowsOptionsTest.cs]: Nit: use `var` here?

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryClientImpl.Queries.cs (hunk 1 above)
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryClientImpl.Queries.cs]: Thanks. I've clearly got something adding this in the way that I do things - it happens far too often :(

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs (hunk 2 above)
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs]: Done. (This is the result of tinkering!)

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs [hunk 4]
```diff
@@ -201,6 +189,23 @@ public BigQueryJob ThrowOnAnyError()
...
 
+        private TableReference GetQueryTableReference()
+        {
+            var query = Resource?.Configuration?.Query;
+            if (query == null)
+            {
+                throw new InvalidOperationException("Job doesn't represent a query");
+            }
+            ThrowOnAnyError();
+            var reference =query.DestinationTable;
```
Comment (evildour) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs]: Nit: add space

Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs]: Whoops, thanks - will fix. I won't merge this until I've talked with the BQ team and ideally DPEs. Thanks for the review though :)

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/QueryTest.cs [hunk 5]
```diff
@@ -334,10 +331,101 @@ public void QueryOnView()
...
+            var queryOptions = new QueryOptions { UseLegacySql = true, UseQueryCache = false };
+            var resultsOptions = new GetQueryResultsOptions { Timeout = TimeSpan.FromSeconds(2) };
+            Assert.Throws<TimeoutException>(() => client.ExecuteQuery(sql, queryOptions, resultsOptions));                
+        }
+
+        [Fact]
+        public void GetQueryResults_Timeout()
+        {
+            // SQL that I happen to know takes over 10 seconds to query.
+            string sql = "SELECT id FROM [bigquery-public-data:github_repos.contents] where content contains 'NodaTime' AND content contains '2.0.2' LIMIT 1000";
```
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/QueryTest.cs]: Have added a commit to skip these tests by default. We can still run them when we want to.

**Comment by jskeet** [comment]
Assigned Mike as he's already reviewed it :)
Just waiting on BQ team before merging.

**Comment by chrisdunelm** [comment]
Great. I'm very happy not to review it ;)
